### DISCUSSION TITLE
Unable to use `async fn` in FromRequestParts implementation

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#2551] [Q&A] [answered]
### Summary

As the async traits feature has been stabilised in rust 1.75, I thought that I would not need the `async_trait` macro from the `async_trait` crate, but when I try to use `async fn` in a trait `FromRequestParts` implementation:

```rust
impl<S: Send + Sync> FromRequestParts<S> for Ctx {
	type Rejection = Error;

	async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self> {
		todo!()
	}
}
```

I get the following error, when I try to compile the provided code: 
lifetime parameters or bounds on method `from_request_parts` do not match the trait declaration

### axum version

0.7.1

**Comment by jplatte** [comment, marked as answer]
This is expected. Switching to native async-fn-in-trait is a breaking change in the library code that will be part of the 0.8.0 release.

**Reply by ghost** [reply]
Thanks for the reply, but I am still confused, if the trait definition does not include async-fn-in-trait fn definition, does it mean, that when implementing the same trait, we would not be able to implement it as async trait, and we would also have to use  `-> impl Future<Output = ...>` or the async_trait macro?? if so, please provide a little bit information about why... because as per my knowledge, the async fn just desugars to `-> impl Future<Output = ...>`, so it should work... I don't know if I'm wrong...

Also can you please explain a bit about the error that I am receiving...

**Reply by jplatte** [reply]
What `async_trait` transforms a trait definition or implementation to is a lot different to what async-fn-in-trait (AFIT) / return-position-impl-trait-in-traits (RPITIT) is, they are just not compatible. Either both the definition and impls need to use `#[async_trait]`, or not. For 0.8, axum's trait definitions will use RPITIT, which *is* compatible with AFIT in `impl`s. For more details, see [the recent rust-lang.org blog post](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html).
